### PR TITLE
SAK-42017: Too much time while loading gradebookNG

### PR DESCRIPTION
**Pull request by jesusmmp**
I would like someone more test this one before merge. Thanks!

### CODE REVIEW

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 1]
```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
-							if (categoryOk && groupOk) {
-								gradeable = true;
-								break;
-							}
 						}
 
 						// set the gradeable flag on this grade instance
 						final GbGradeInfo gradeInfo = entry.getValue();
-						gradeInfo.setGradeable(gradeable);
+						gradeInfo.setGradeable(gradeable||allPermissions);
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Spaces! All small things, just to make it more readable.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 2]
```diff
@@ -1311,6 +1312,25 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
 				// get the group membership for the students
 				final Map<String, List<String>> groupMembershipsMap = getGroupMemberships();
 
+				//Pair group <-> category
+				Map <Long, List<String>> permByCat = new HashMap<>();
+
+				for (final PermissionDefinition permission : permissions) {
```
[the remark this answers is not in the document]
Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: It was this way before, I've just keep it final

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 3]
```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
 							}
 
-							// if group reference is null (can grade all groups) or group membership contains student (can grade this group)
-							if (StringUtils.isBlank(permissionGroupReference)) {
-								groupOk = true;
-								log.debug("Group check passed #1");
-							} else {
-								final List<String> groupMembers = groupMembershipsMap.get(permissionGroupReference);
-								log.debug("groupMembers: {}", groupMembers);
+							for (String group:(List<String>) arr) {
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: You need to space this out as it is hard to read. Also, you don't need  to cast arr to a List<String> - it's already that.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 4]
```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
-
-							boolean categoryOk = false;
-							boolean groupOk = false;
+						//If TA does not have all the permissions over categories and groups
+						if (!allPermissions) {
 
-							final Long permissionCategoryId = permission.getCategoryId();
-							final String permissionGroupReference = permission.getGroupReference();
+							//Check category and group permissions
+							List<String> arr = permByCat.get(gradeCategoryId) != null ? permByCat.get(gradeCategoryId) : new ArrayList<String>();
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: You're doing two map lookups here when you don't need to.

List\<String\> arr = permByCat.getOrDefault(gradeCategoryId, new ArrayList\<String\>());

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 5]
```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
 
-							log.debug("permissionCategoryId: {}", permissionCategoryId);
-							log.debug("permissionGroupReference: {}", permissionGroupReference);
-
-							// if permissions category is null (can grade all categories) or they match (can grade this category)
-							if (!categoriesAreEnabled() || (permissionCategoryId == null || permissionCategoryId.equals(gradeCategoryId))) {
-								categoryOk = true;
-								log.debug("Category check passed");
+							//add groups with permission in all categories
+							if (permByCat.containsKey(Long.valueOf(-1L))) {
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: You're doing two map searches here  when you don't need to. You could just go:
arr.addAll(permByCat.getOrDefault(Long.valueOf(-1L), Collections.\<String\>emptyList());

or ...

Optional.ofNullable(permByCat.get(Long.valueOf(-1L))).ifPresent(arr::addAll)

from https://stackoverflow.com/questions/30739105/any-null-safe-alternative-to-arraylist-addall

I like the second one, tbh.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 6]
```diff
@@ -1311,6 +1312,25 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
 				// get the group membership for the students
 				final Map<String, List<String>> groupMembershipsMap = getGroupMemberships();
 
+				//Pair group <-> category
+				Map <Long, List<String>> permByCat = new HashMap<>();
+
+				for (final PermissionDefinition permission : permissions) {
+					final Long permissionCategoryId = permission.getCategoryId() !=null ? permission.getCategoryId() : -1L;
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: != null

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 7]
```diff
@@ -1311,6 +1312,25 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
+					final String permissionGroupReference = permission.getGroupReference() !=null ? permission.getGroupReference() : "all";
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: != null

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 8]
```diff
@@ -1311,6 +1312,25 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
...
+
+					//permissions over all categories and grades
+					if (Long.valueOf(-1L).equals(permissionCategoryId) && "all".equals(permissionGroupReference)) {
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Where does "all" come from? Could it be in a final String somewhere, a constant?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 7 above)
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: "all" is a bit magic :)

**Comment by jesusmmp** [comment]
I've just updated all issues. Thanks @adrianfish !

**Pull request merged**